Declarative marking of RTypedData struct references

eightbitraptor opened the pull request:
This PR is a work in progress.

I've added a flag `RUBY_TYPED_DECL_MARKING` that can be used inside the `rb_data_type_t` flags section to determine whether a T_DATA object can be declaratively marked or not.

If this flag is set then we can use the `data` struct member as a pointer to an array of offsets that we can use in conjunction with the `data` struct to find fields that require marking.

I've modified `gc_mark_children` to use this offset array if present to test the system.

eightbitraptor committed:
wip declarative marking of RTypedData references

chrisseaton commented on dir.c:
```diff
@@ -487,10 +487,14 @@ dir_memsize(const void *ptr)
 
+#define END_REFS UINTPTR_MAX
+
+static VALUE references[] = {offsetof(struct dir_data, path), END_REFS};
```
This should be `offset_t references[]` shouldn't it?

And, after it works, we should then transition it to be a bitmap.

eightbitraptor replied:
I think this should probably be `size_t references[]` (`off_t` is part of POSIX, and the docs tell me it's normally used for file sizes).

I assume I didn't get warned about this is that both VALUE and `size_t` are the same size integer types...

chrisseaton replied:
Looks like you're right they should be `size_t`.

chrisseaton commented:
This is the general idea as I understood it, yes.

tompng committed:
[ruby/reline] Fix dialog scrollbar rendering position and disappearing bug

https://github.com/ruby/reline/commit/e21b69ade4

tompng committed:
[ruby/reline] Add scrollbar scroll-to-bottom test and fix existing scrollbar rendeing test

https://github.com/ruby/reline/commit/268e02b155

tompng committed:
[ruby/reline] Add constant MINIMUM_SCROLLBAR_HEIGHT for scrollbar rendering when dialog content is very long

https://github.com/ruby/reline/commit/d4c95f89ec

peterzhu2118 committed:
Fix crash when defining ivars on special constants

[Bug #19339]

peterzhu2118 committed:
Don't redefine RB_OBJ_WRITE

RB_OBJ_WRITE already exists in rgengc.h, so we shouldn't redefine it in
gc.h.

st0012 committed:
Avoid checking interrupt when loading iseq

The interrupt check will unintentionally release the VM lock when loading an iseq.
And this will cause issues with the `debug` gem's
[`ObjectSpace.each_iseq` method](https://github.com/ruby/debug/blob/0fcfc28acae33ec1c08068fb7c33703cfa681fa7/ext/debug/iseq_collector.c#L61-L67),
which wraps iseqs with a wrapper and exposes their internal states when they're actually not ready to be used.

And when that happens, errors like this would occur and kill the `debug` gem's thread:

```
 DEBUGGER: ReaderThreadError: uninitialized InstructionSequence
┃ DEBUGGER: Disconnected.
┃ ["/opt/rubies/ruby-3.2.0/lib/ruby/gems/3.2.0/gems/debug-1.7.1/lib/debug/breakpoint.rb:247:in `absolute_path'",
┃  "/opt/rubies/ruby-3.2.0/lib/ruby/gems/3.2.0/gems/debug-1.7.1/lib/debug/breakpoint.rb:247:in `block in iterate_iseq'",
┃  "/opt/rubies/ruby-3.2.0/lib/ruby/gems/3.2.0/gems/debug-1.7.1/lib/debug/breakpoint.rb:246:in `each_iseq'",
...
```

A way to reproduce the issue is to satisfy these conditions at the same time:

1. `debug` gem calling `ObjectSpace.each_iseq` (e.g. [activating a `LineBreakpoint`](https://github.com/ruby/debug/blob/0fcfc28acae33ec1c08068fb7c33703cfa681fa7/lib/debug/breakpoint.rb#L246)).
2. A large amount of iseq being loaded from another thread (possibly through the `bootsnap` gem).
3. 1 and 2 iterating through the same iseq(s) at the same time.

Because this issue requires external dependencies and a rather complicated timing setup to reproduce, I wasn't able to write a test case for it.
But here's some pseudo code to help reproduce it:

```rb
require "debug/session"

Thread.new do
  100.times do
    ObjectSpace.each_iseq do |iseq|
      iseq.absolute_path
    end
  end
end

sleep 0.1

load_a_bunch_of_iseq
possibly_through_bootsnap
```

[Bug #19348]

Co-authored-by: Peter Zhu <peter@peterzhu.ca>

dependabot[bot] committed:
[rubygems/rubygems] Bump rb-sys

Bumps [rb-sys](https://github.com/oxidize-rb/rb-sys) from 0.9.54 to 0.9.56.
- [Release notes](https://github.com/oxidize-rb/rb-sys/releases)
- [Commits](https://github.com/oxidize-rb/rb-sys/compare/v0.9.54...v0.9.56)

---
updated-dependencies:
- dependency-name: rb-sys
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

hsbt committed:
SimpleCov is hosted under simplecov-ruby org now

hsbt committed:
Switch to use gem version of simplecov, not git clone

hsbt committed:
Pin simplecov-0.20.0

dependabot[bot] committed:
[rubygems/rubygems] Bump rb-sys in /test/rubygems/test_gem_ext_cargo_builder/custom_name

Bumps [rb-sys](https://github.com/oxidize-rb/rb-sys) from 0.9.54 to 0.9.56.
- [Release notes](https://github.com/oxidize-rb/rb-sys/releases)
- [Commits](https://github.com/oxidize-rb/rb-sys/compare/v0.9.54...v0.9.56)

---
updated-dependencies:
- dependency-name: rb-sys
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

hsbt committed:
Added test-syntax-suggest to help message

hsbt committed:
Added make exam to help message

hsbt committed:
test-syntax-suggest is now part of make check

hsbt committed:
[ruby/psych] configure of libyaml couldn't detect "arm64-apple-darwin22" for build host.

  checking whether make sets $(MAKE)... (cached) yes
  checking build system type... arm-apple-darwin21.6.0
  checking host system type... Invalid configuration `arm64-apple-darwin21': machine `arm64-apple' not recognized
  configure: error: /bin/sh /Users/hsbt/Downloads/yaml-0.2.5/config/config.sub arm64-apple-darwin21 failed
  *** extconf.rb failed ***

https://github.com/ruby/psych/commit/ad1502202c

hsbt committed:
[ruby/psych] Bump up 5.0.2.pre1 for testing

https://github.com/ruby/psych/commit/67ec299c68

matzbot committed:
Update default gems list at 492d37fbbd3613b7fb180eec529860 [ci skip]

peterzhu2118 committed:
Remove macro RHASH_ITER_LEV

The function rb_hash_iter_lev doesn't exist as it was removed.

Commit:
Update yjit.md

nobu committed:
Define RUBY_VERSION_IS_3_3 macro in rubyspec.h

znz committed:
Remove removed files from ext/.document [ci skip]

- dbm/dbm.c removed at 55cd3e4ebff8fa75854ecadcd77abbf7cf4b5823
- fiber/fiber.c removed at 521ad9a13aef1e1dae4eaff70d3e6b2e358aa095
- gdbm/gdbm.c removed at edcc29dcff1b269b7748ab83adf21b2f3f97ebff

znz committed:
Clean up removed directories in ext/Setup* [ci skip]

hsbt committed:
[ruby/psych] Bump up 5.0.2

https://github.com/ruby/psych/commit/a170b8eb46

matzbot committed:
Update default gems list at 8a771efec21c9dacad5bb8bf14b81c [ci skip]

kaiquekandykoga committed:
Change ArgumentError message when Comparable#clamp receives min value higher than max value

tompng committed:
[ruby/reline] multiline_repl do not need to depend on RubyLex
(https://github.com/ruby/reline/pull/502)

* multiline_repl do not need to depend on RubyLex

* Add auto indent test

soutaro committed:
Update RBS commit hash (#7143)

Use a commit at the top of RBS 2.8.3, not 3.0.0.dev.N.

k0kubun committed:
benchmark_driver v0.16.3

Some chruby-related features and bug fixes.

hsbt committed:
Supressing warnings messages like:

  Ignoring debug-1.7.1 because its extensions are not built. Try: gem pristine debug --version 1.7.

  Revert "Clear gem paths for each test"

  This reverts commit 6698b580ddad8cfa8c5c86df9328472820d3ee6a.

nobu committed:
Make installation messages verbose a little [ci skip]

Commit:
Add stats so we can keep track of x86 rel32 vs register calls (#7142)

* Add stats so we can keep track of x86 rel32 vs register calls

To know if we get that "prime real estate" as Alan put it.

* Fix bug pointed by Alan

Commit:
YJIT: implement codegen for `String#empty?` (#7148)

YJIT: implement codegen for String#empty?

BurdetteLamar committed:
[ruby/net-http] [DOC] Enhanced RDoc for set_form
(https://github.com/ruby/net-http/pull/103)

https://github.com/ruby/net-http/commit/f6506ff889

ioquatix committed:
Test some missing coverage too. (#7041)

XrXr committed:
YJIT: Use .as_side_exit() for jumps to counted exits

Fewer cycles running nops when these jumps are not taken. Fixing all
these so when they get copy pasted in the future we save on padding.

k0kubun committed:
YJIT: Set RUST_BACKTRACE=1 on YJIT GitHub Actions (#7152)

shugo committed:
String#bytesplice should return self

In Feature #19314, we concluded that the return value of String#bytesplice
should be changed from the source string to the receiver, because the source
string is useless and confusing when extra arguments are added.

This change should be included in Ruby 3.2.1.

eightbitraptor commented:
I'm going to investigate and try and benchmark this approach on top of Ruby master, so I'm closing this PR for now to avoid having a convoluted history on this fork.
